### PR TITLE
Target base scale support

### PR DESCRIPTION
**Pull request by alrod**
https://github.com/Azure/azure-functions-host/pull/8780
https://github.com/Azure/azure-sdk-for-net/pull/31367

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScaleStatusContext.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        /// </summary>
+        /// <remarks>
+        /// Value is reolved in the <see cref="ITargetScaler"/> implementation if null.
+        /// </remarks>
+        public int? InstanceConcurrency { get; set; }
+
+        /// <summary>
+        /// The current worker count for the host application.
+        /// </summary>
+        public int WorkerCount { get; set; }
```
Comment (alrod) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScaleStatusContext.cs]: In initial design we did not want `WorkerCount` but it is needed as we are calculating number of added/removed workers.

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerResult.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+{
+    /// <summary>
+    /// Represents result for target base scale.
+    /// </summary>
+    public class TargetScalerResult
+    {
+        /// <summary>
+        /// Gets or sets the target worker count.
+        /// </summary>
+        public int WorkerCountDifference { get; set; }
```
Comment (alrod) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerResult.cs]: In initial design we wanted `TargetWorkerCount` but actually we need the difference as "admin/host/scale/status" also is used by "regular" scale.

File: src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs [hunk 3]
```diff
@@ -70,7 +74,19 @@ public async Task<IListener> CreateAsync(CancellationToken cancellationToken)
 
                 IListener listener = await listenerFactory.CreateAsync(cancellationToken);
 
-                RegisterScaleMonitor(listener, _monitorManager);
+                bool targetScalerRegistered = false;
+                if (bool.TryParse(Environment.GetEnvironmentVariable(TargetBaseScalingEnabled), out bool isTargetBaseScalingEnabled))
```
Comment (chiangvincent) [inline src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs]: In addition to the `TARGET_BASED_SCALING_ENABLED` app setting, we also need to check if TBSA is enabled (per extension type) via the `ScaleControllerFeatureFlags` hosting configuration. See [Scale Controller PR](https://msazure.visualstudio.com/One/_git/AAPT-Antares-ScaleController/pullrequest/6552084) and [TBS +RDS design doc](https://microsoft-my.sharepoint.com/:w:/p/vchiang/Edf-XUIR7ZpHuwhZYporzlUB3UkSnjPMV9Wz94UPU4bV4g?e=9FwGKq) for reference

File: src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScaler.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Interface defining a target scaler that can participate in Azure Functions scaling decisions by
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScaler.cs]: "Interface defining an Azure Functions scaler that makes scale decisions based on current event source metrics and function concurrency."

File: src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    /// <summary>
+    /// Manager for registering and accessing <see cref="ITargetScaler"/> instances for
+    /// a <see cref="JobHost"/> instance.
+    /// </summary>
+    public interface ITargetScalerManager
+    {
+        /// <summary>
+        /// Register an <see cref="ITargetScaler"/> instance.
+        /// </summary>
+        /// <param name="monitor">The monitor instance to register.</param>
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs]: The scaler instance to register

File: src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        void Register(ITargetScaler monitor);
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs]: scaler

File: src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+        /// <summary>
+        /// Get all registered target scale instances.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs]: target scaler instances

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    public class TargetScalerDescriptor
+    {
+        public TargetScalerDescriptor(string id, string functionId)
+        {
+            Id = id;
+            FunctionId = functionId;
+        }
+
+        /// <summary>
+        /// Gets the unique ID for the monitor.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs]: You're referencing monitors here which doesn't apply. I don't think we need this ID property at all. Monitors needed a fully formed ID for persistence, which we don't need for target scalers?

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        /// with their corresponding monitor instance. E.g. for a QueueTrigger, this might
+        /// be of the form "{FunctionId}-QueueTrigger-{QueueName}".
+        /// </remarks>
+        public string Id { get; }
+
+        /// <summary>
+        /// Gets the unique function id.
+        /// </summary>
+        /// <remarks>
+        /// This should be constant. It is used to get persisted dynamic concurrency value.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs]: Let's remove this remark.

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        /// </summary>
+        /// <remarks>
+        /// This should be constant. It is used to correlate persisted metrics samples
+        /// with their corresponding monitor instance. E.g. for a QueueTrigger, this might
+        /// be of the form "{FunctionId}-QueueTrigger-{QueueName}".
+        /// </remarks>
+        public string Id { get; }
+
+        /// <summary>
+        /// Gets the unique function id.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs]: "Gets the ID of the function associated with this scaler."

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerResult.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Represents result for target base scale.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerResult.cs]: "Represents the scale result of a target scaler."

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerResult.cs (hunk 2 above)
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerResult.cs]: I don't think scalers should be returning this - instead they should just return a TargetWorkerCount.

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerStatusContext.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        /// </summary>
+        /// <remarks>
+        /// Value is reolved in the <see cref="ITargetScaler"/> implementation if null.
+        /// </remarks>
+        public int? InstanceConcurrency { get; set; }
+
+        /// <summary>
+        /// The current worker count for the host application.
+        /// </summary>
+        public int WorkerCount { get; set; }
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerStatusContext.cs]: WorkerCount isn't needed for target scaling - I think we can remove this

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerStatusContext.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Context used by <see cref="ITargetScaler.GetScaleResultAsync(TargetScalerStatusContext)"/> to decide
+    /// scale result.
+    /// </summary>
+    public class TargetScalerStatusContext
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerStatusContext.cs]: Rename to TargetScalerContext?

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerStatusContext.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        /// <summary>
+        /// The current worker dyanimc worker concurrency.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerStatusContext.cs]: "The per instance concurrency for the associated function"

In remarks we can say "When not specified, the scaler will determine the concurrency based on configuration."

File: test/Microsoft.Azure.WebJobs.Host.FunctionalTests/HostListenerFactoryTests.cs [hunk 15]
```diff
@@ -37,36 +38,70 @@ public HostListenerFactoryTests()
...
+        [InlineData(false, typeof(TestListener_Monitor), typeof(TestListener_TargetScaler), 1, 0)]
+        [InlineData(false, typeof(TestListener_Monitor), typeof(TestListener_TargetScalerProvider), 1, 0)]
+        [InlineData(false, typeof(TestListener_TargetScaler), typeof(TestListener_TargetScalerProvider), 0, 0)]
+        [InlineData(false, typeof(TestListener_Monitor), typeof(TestListener_MonitorAndTargetScaler), 2, 0)]
+        [InlineData(true, typeof(TestListener_Monitor), typeof(TestListener_Monitor), 2, 0)]
+        [InlineData(true, typeof(TestListener_Monitor), typeof(TestListener_TargetScaler), 1, 1)]
+        [InlineData(true, typeof(TestListener_Monitor), typeof(TestListener_TargetScalerProvider), 1, 1)]
+        [InlineData(true, typeof(TestListener_TargetScaler), typeof(TestListener_TargetScalerProvider), 0, 2)]
+        [InlineData(true, typeof(TestListener_Monitor), typeof(TestListener_MonitorAndTargetScaler), 1, 1)]
+        public async Task CreateAsync_RegistersScaleMonitorsAndTargetScaler(bool tbsEnabled,
```
Comment (mathewc) [inline test/Microsoft.Azure.WebJobs.Host.FunctionalTests/HostListenerFactoryTests.cs]: nit: CreateAsync_RegistersScaleMonitorsAndTargetScalers

File: src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs [hunk 16]
```diff
@@ -14,13 +14,15 @@
...
 using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
 
 namespace Microsoft.Azure.WebJobs.Host.Listeners
 {
     internal class HostListenerFactory : IListenerFactory
     {
         private static readonly MethodInfo JobActivatorCreateMethod = typeof(IJobActivator).GetMethod("CreateInstance", BindingFlags.Public | BindingFlags.Instance).GetGenericMethodDefinition();
         private const string IsDisabledFunctionName = "IsDisabled";
+        private const string TargetBaseScalingEnabled = "TARGET_BASED_SCALING_ENABLED";
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs]: Let's add a constant for this and use it in the source/tests

File: src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs (hunk 3 above)
Comment (mathewc) [reply src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs]: I wonder if we shouldn't just remove this env var check from the SDK. SDK should just register all monitors and scalers, and it's up to the consumer (Functions Host) to determine whether based on environment they should be used. Note that's how monitors already work - SDK registers them, but Functions Host only uses them based on FUNCTIONS_RUNTIME_SCALE_MONITORING_ENABLED. This keeps the SDK clean w/o any Functions specific env settings, and allows the functions host to decide which to use.

**Comment by seilorjunior** [comment]
@seilorjunior

File: src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs (hunk 3 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs]: Moved both checks (TARGET_BASED_SCALING_ENABLED  and hosting configuration per individual trigger) to function host

File: src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs (hunk 16 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs]: This check was moved to function host

File: test/Microsoft.Azure.WebJobs.Host.FunctionalTests/HostListenerFactoryTests.cs (hunk 15 above)
Comment (alrod) [reply test/Microsoft.Azure.WebJobs.Host.FunctionalTests/HostListenerFactoryTests.cs]: The test is not needed anymore as we are registering everything and function host decides what to use

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs (hunk 8 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs:19]: I am using the "ID" to "merge" scale monitors and target scalers in function host as we register everything in webjobs sdk - if target scaler is defined and TBS is enabled we do not need to use scale monitor.

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerContext.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Context used by <see cref="ITargetScaler.GetScaleResultAsync(TargetScalerContext)"/> to decide
+    /// scale result.
+    /// </summary>
+    public class TargetScalerContext
+    {
+        /// <summary>
+        /// The current worker dyanimc worker concurrency.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerContext.cs]: nit: The current concurrency for the target function.

File: src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs [hunk 18]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/TargetScalerDescriptor.cs]: nit: extra space

File: src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+        void Register(ITargetScaler scaler);
+
+        /// <summary>
+        /// Get all registered target scaler instances.
+        /// </summary>
+        /// <remarks>
+        /// Should only be called after the host has been started and all
+        /// instances are registered.
+        /// </remarks>
+        /// <returns>The collection of target scaler intances.</returns>
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Scale/ITargetScalerManager.cs]: spelling

**Review by mathewc — APPROVED** [review]

**Review by chiangvincent — APPROVED** [review]

**Pull request merged**